src/repr/src/row.rs

```diff
@@ -121,24 +121,21 @@ impl Row {
         proto: &ProtoRow,
         desc: &RelationDesc,
     ) -> Result<(), String> {
-        let mut col_idx = 0;
         let mut packer = self.packer();
-        for d in proto.datums.iter() {
+        for (col_idx, _, _) in desc.iter_all() {
+            let d = match proto.datums.get(col_idx.to_raw()) {
+                Some(x) => x,
+                None => {
+                    packer.push(Datum::Null);
+                    continue;
+                }
+            };
             packer.try_push_proto(d)?;
-            col_idx += 1;
         }
 
         let num_columns = desc.typ().column_types.len();
-        if col_idx < num_columns {
-            let missing_columns = col_idx..num_columns;
-            for _ in missing_columns {
-                packer.push(Datum::Null);
-                col_idx += 1;
-            }
-        }
-
         mz_ore::soft_assert_eq_or_log!(
-            col_idx,
```

Comment:
A little more expensive in the new version, since it involves some decoding... unsure if it crosses the caring threshold.

(This strikes me as a less interesting assert post-refactoring... we're iterating over the desc instead of the input now, so it would be very hard to end up with a different number of cols.)

Reply:
Yeah good call, the check doesn't seem that interesting to me either so I removed it